operators/audio/audio_decoder.hpp

```diff
@@ -0,0 +1,157 @@
+#include "ocos.h"
+
+#include <list>
+#define DR_FLAC_IMPLEMENTATION
+#include "dr_flac.h"
+#define DR_MP3_IMPLEMENTATION 1
+#define DR_MP3_FLOAT_OUTPUT 1
+#include "dr_mp3.h"
+#define DR_WAV_IMPLEMENTATION
+#include "dr_wav.h"
+
+#include "string_utils.h"
+#include "string_tensor.h"
+
+struct KernelAudioDecoder : public BaseKernel {
+ public:
+  KernelAudioDecoder(const OrtApi& api, const OrtKernelInfo& info) : BaseKernel(api, info) {
+  }
+
+  enum {
+    DEFAULT_STREAM = 0,
+    WAV_STREAM,
+    MP3_STREAM,
+    FLAC_STREAM
+  };
+
+  int ReadStreamFormat(OrtKernelContext* context, const int64_t* p_data) {
+    int stream_format = DEFAULT_STREAM;
+    const OrtValue* ov_format = ort_.KernelContext_GetInput(context, 1);
+    if (ov_format != nullptr) {
+      std::vector<std::string> str_format;
+      GetTensorMutableDataString(api_, ort_, context, ov_format, str_format);
+      const char* formats[] = {"default",
+                               "wav",
+                               "mp3",
+                               "flac"};
+      for (auto fmt : formats) {
+        if (str_format[0] == fmt) {
+          break;
+        }
+        stream_format++;
+      }
+    }
+
+    if (stream_format == DEFAULT_STREAM){
+      auto p_stream = reinterpret_cast<char const*>(p_data);
+      std::string_view marker(p_stream, 4);
+      if (marker == "fLaC") {
+        stream_format = FLAC_STREAM;
+      } else if (marker == "RIFF") {
+        stream_format = WAV_STREAM;
+      } else if (marker[0] == char(-1) && marker[1] == char(-13)) {
+        stream_format = MP3_STREAM;
+      } else {
+        ORTX_CXX_API_THROW("[AudioDecoder]: Unknown audio stream format", ORT_INVALID_ARGUMENT);
+      }
+    }
+    return stream_format;
+  }
+
+  template<typename _CLS_AUDIO, typename _FN_DECODER>
+  static size_t DrReadFrames(std::list<std::vector<float>>& frames, _FN_DECODER fx, _CLS_AUDIO& obj) {
+    const size_t default_chunk_size = 4096;
+    int64_t total_frames = 0;
+
+    for (;;) {
+      std::vector<float> buf;
+      buf.resize(default_chunk_size * obj.channels);
+      auto n_frames = fx(&obj, default_chunk_size, buf.data());
+      if (n_frames <= 0) {
+        break;
+      }
+      auto n_samples = n_frames * obj.channels;
+      total_frames += n_samples;
+      buf.resize(n_samples);
+      frames.emplace_back(buf);
+    }
+
+    return total_frames;
+  }
+
+  void Compute(OrtKernelContext* context) {
+    const OrtValue* input = ort_.KernelContext_GetInput(context, 0);
+    const int64_t* p_data = ort_.GetTensorData<int64_t>(input);
+
+    OrtTensorDimensions input_dim(ort_, input);
+    if (!((input_dim.size() == 1) || (input_dim.size() == 2 && input_dim[0] == 1))) {
+      ORTX_CXX_API_THROW("[AudioDecoder]: Expect input dimension [n] or [1,n].", ORT_INVALID_ARGUMENT);
+    }
+
+    auto stream_format = ReadStreamFormat(context, p_data);
+
+    int64_t total_frames = 0;
+    std::list<std::vector<float>> lst_frames;
```

Comment:
why `std::list`? I think generally `std::vector` is preferable.

Reply:
actually, since we just copy each element (float vector) to the output, could we just use a single float vector to store all frame data?

Reply:
because we don't know the buf size of final decoding stream, not using a single float vector is to avoid the big memory allocation overhead. so does list vs. vector.

Reply:
ah, I see.

is it possible to query the frame/sample (I don't know the right word) size of the whole stream beforehand and avoid repeated dynamic memory allocations and copies?

e.g., with APIs like these:
https://github.com/mackron/dr_libs/blob/4b3d07849537ce0b71b22180c0b1335eacc6e9be/dr_mp3.h#L440-L458

Reply:
mp3 doesn't have a simple and light-weight to cal the final decompressed data size, like reading the value from the header. And these get_frame_count APIs needs decoding the frame firstly which makes it look heavy. So, it is not unworthy of adding a dependency on these APIs.

I have increased the decoding buf size to 256K from 4K(which seems to be small), the memory allocation should be less frequent and should not be a problem now.

Reply:
ok. I guess this is fine for the initial implementation, but I think there might be some potential performance improvement here.